File: example projects/example-ruby/validator.rb
require 'digest'

module App
  module Validator
    EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
    USERNAME_REGEX = /\A[a-zA-Z0-9_]+\z/

    def self.validate_email(email)
      return false if email.nil? || email.empty?
      !!(email =~ EMAIL_REGEX)
    end

    def self.hash_password(password)
      raise ArgumentError, 'Password must be at least 8 characters' if password.length < 8
      Digest::SHA256.hexdigest(password)
    end

    def self.verify_password(password, hash)
      Digest::SHA256.hexdigest(password) == hash
    end

    def self.sanitize_string(input)
      return '' if input.nil? || input.empty?

      sanitized = input.strip
      sanitized.gsub(/[<>]/, '')
    end

    def self.validate_username(username)
      return false if username.nil? || username.empty?
      return false if username.length < 3 || username.length > 20
      !!(username =~ USERNAME_REGEX)
    end

    def self.validate_password_strength(password)
      return false if password.length < 8

      has_uppercase = password =~ /[A-Z]/
      has_lowercase = password =~ /[a-z]/
      has_digit = password =~ /[0-9]/
      has_special = password =~ /[!@#$%^&*(),.?":{}|<>]/

      [has_uppercase, has_lowercase, has_digit, has_special].count(true) >= 3
    end
  end
end

File: example projects/example-vb/Validator.vb
Imports System.Text.RegularExpressions
Imports System.Security.Cryptography
Imports System.Text

Namespace App.Utils
    Public Class Validator
        Private Shared ReadOnly EmailRegex As New Regex("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$")
        Private Shared ReadOnly UsernameRegex As New Regex("^[a-zA-Z0-9_]+$")
        
        Public Shared Function ValidateEmail(email As String) As Boolean
            If String.IsNullOrEmpty(email) Then
                Return False
            End If
            
            Return EmailRegex.IsMatch(email)
        End Function
        
        Public Shared Function HashPassword(password As String) As String
            If password.Length < 8 Then
                Throw New ArgumentException("Password must be at least 8 characters")
            End If
            
            Using sha256 As SHA256 = SHA256.Create()
                Dim bytes As Byte() = Encoding.UTF8.GetBytes(password)
                Dim hash As Byte() = sha256.ComputeHash(bytes)
                
                Dim builder As New StringBuilder()
                For Each b As Byte In hash
                    builder.Append(b.ToString("x2"))
                Next
                
                Return builder.ToString()
            End Using
        End Function
        
        Public Shared Function VerifyPassword(password As String, hash As String) As Boolean
            Try
                Dim computedHash As String = HashPassword(password)
                Return computedHash = hash
            Catch
                Return False
            End Try
        End Function
        
        Public Shared Function SanitizeString(input As String) As String
            If String.IsNullOrEmpty(input) Then
                Return String.Empty
            End If
            
            Dim sanitized As String = input.Trim()
            sanitized = sanitized.Replace("<", "&lt;")
            sanitized = sanitized.Replace(">", "&gt;")
            sanitized = sanitized.Replace("""", "&quot;")
            sanitized = sanitized.Replace("'", "&#x27;")
            
            Return sanitized
        End Function
        
        Public Shared Function ValidateUsername(username As String) As Boolean
            If String.IsNullOrEmpty(username) Then
                Return False
            End If
            
            If username.Length < 3 OrElse username.Length > 20 Then
                Return False
            End If
            
            Return UsernameRegex.IsMatch(username)
        End Function
        
        Public Shared Function ValidatePasswordStrength(password As String) As Boolean
            If password.Length < 8 Then
                Return False
            End If
            
            Dim hasUppercase As Boolean = password.Any(AddressOf Char.IsUpper)
            Dim hasLowercase As Boolean = password.Any(AddressOf Char.IsLower)
            Dim hasDigit As Boolean = password.Any(AddressOf Char.IsDigit)
            Dim hasSpecial As Boolean = password.Any(Function(c) Not Char.IsLetterOrDigit(c))
            
            Dim criteriaCount As Integer = 0
            If hasUppercase Then criteriaCount += 1
            If hasLowercase Then criteriaCount += 1
            If hasDigit Then criteriaCount += 1
            If hasSpecial Then criteriaCount += 1
            
            Return criteriaCount >= 3
        End Function
    End Class
End Namespace
